### PR TITLE
OpenGL updates

### PR DESCRIPTION
**Pull request by petersteneteg**

### CODE REVIEW

File: include/inviwo/core/util/moveonlyvalue.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+#pragma once
+
+#include <optional>
+#include <utility>
+
+namespace inviwo {
+
+/**
+ * @brief Hold a "unique" value, the value will not be copied and when mowing the moved from
+ * optional will be left empty
```
Comment (ResearchDaniel) [inline include/inviwo/core/util/moveonlyvalue.h:38]: Typo, moving

File: modules/opengl/include/modules/opengl/buffer/framebufferobject.h [hunk 2]
```diff
@@ -50,163 +60,296 @@ class IVW_MODULE_OPENGL_API FrameBufferObject {
...
-     * Use this function to unbind the FBO, without restoring the previous after this call, no FBO
-     * is bound
+     * @brief Unbind the framebuffer (binds id 0)
      */
     static void deactivateFBO();
 
     /**
-     *	For attaching a 2D Texture
+     * @brief Check it this framebuffer is currently bound
+     */
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/buffer/framebufferobject.h:84]: Typo, if

File: modules/opengl/include/modules/opengl/buffer/framebufferobject.h [hunk 3]
```diff
@@ -50,163 +60,296 @@ class IVW_MODULE_OPENGL_API FrameBufferObject {
...
+    bool isActive() const;
+
+    /**
+     * @brief Check the framebuffer status. Log a warning if it is not Complete.
+     * @return true if the framebuffer is complete otherwise false.
+     */
+    bool checkStatus() const;
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/buffer/framebufferobject.h]: Should there be an option for logging? 
Or separate into two. One for checking status and one for checking and logging?

File: modules/opengl/include/modules/opengl/buffer/framebufferobject.h [hunk 4]
```diff
@@ -50,163 +60,296 @@ class IVW_MODULE_OPENGL_API FrameBufferObject {
...
-     * location in the draw buffer list (as used for attrib location in shaders). Side effects:
-     * affects subsequent buffer locations of already attached color buffers. NOTE: if
-     * forcedLocation is larger than the number of attachments, it will not be considered
-     */
-    GLuint performAttachColorTexture(GLuint texId, int attachmentNumber,
-                                     bool attachFromRear = false, int forcedLocation = -1);
-
-    std::string printBuffers() const;
-    static std::string getAttachmentStr(GLenum attachmentID);
+    static GLuint maxColorattachments();
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/buffer/framebufferobject.h]: Duplicate of getMaxColorAttachments?
Also, camel case for Attachments

File: modules/opengl/include/modules/opengl/image/imagegl.h [hunk 5]
```diff
@@ -46,27 +45,84 @@ class Image;
...
+     *
+     * Binds the framebuffer, and set the draw buffers. Color layer 0 will be at location 0, if
+     * picking is included in the \p imageType the picking layer will be at location 1. Any
+     * additional layers will be at the following locations. If \p imageType contains depth the
+     * depth test is enabled and the depth mask is set to true, otherwise the depth test is disabled
+     * and the depth mask is set to false. Finally the viewport is the to the image dimensions.
+     * @param imageType The layers that should be active.
+     * @return An RAII object that will restore the old state on destruction.
+     */
+    void activateBuffer(ImageType imageType = ImageType::AllLayers);
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/image/imagegl.h:91]: Should return the state

File: modules/opengl/include/modules/opengl/image/imagegl.h [hunk 6]
```diff
@@ -46,27 +45,84 @@ class Image;
...
+     * @copydoc copyRepresentationsTo(ImageRepresentation* target) const
+     */
     bool copyRepresentationsTo(ImageGL* target) const;
-    bool updateFrom(const ImageGL*);
+
+    /**
+     * @brief Copies @p source into this using Blitting and PBOs if needed.
+     *
+     * Does not care about aspect ratios, copies the source image into the this using "nearest"
+     * interpolation. If some layers are not attached to the FBO they will be copies using PBO
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/image/imagegl.h]: Typo, copied

File: modules/opengl/src/image/imagegl.cpp [hunk 7]
```diff
@@ -45,74 +48,50 @@ ImageGL::ImageGL()
     , depthLayerGL_{nullptr}
     , pickingLayerGL_{nullptr}
     , frameBufferObject_{}
-    , pickingAttachmentID_{0}
-    , colorLayerCopyCount_{0}
-    , prevDepthTest_{false}
-    , prevDepthMask_{false} {}
+    , colorAttachmentIDs_{}
```
Comment (ResearchDaniel) [inline modules/opengl/src/image/imagegl.cpp:51]: Could initialize these in the header to avoid duplicating code in the constructors?

**Review by ResearchDaniel — CHANGES_REQUESTED** [review]
Nice! Some minor comments

File: modules/opengl/include/modules/opengl/image/imagegl.h (hunk 5 above)
Comment (petersteneteg) [reply modules/opengl/include/modules/opengl/image/imagegl.h:91]: This one does not save the state only the new active one does that. In most cases it is not needed. But I see i made a copy paste misstake in the doc...

File: modules/opengl/src/image/imagegl.cpp (hunk 7 above)
Comment (petersteneteg) [reply modules/opengl/src/image/imagegl.cpp:51]: Im not too fond of initialising stuff in the header. It leaks implementation details into the header, it means I have to recompile when ever I change something, and it separates the constructions code into two different places. But maybe I will come around at some point.

Comment (ResearchDaniel) [reply modules/opengl/src/image/imagegl.cpp:51]: Yeah, I guess it has its pros and cons. Does not matter too much

File: modules/opengl/include/modules/opengl/image/imagegl.h (hunk 5 above)
Comment (ResearchDaniel) [reply modules/opengl/include/modules/opengl/image/imagegl.h:91]: I see. Perhaps we could make a note on which one should be used an for which cases

File: modules/opengl/include/modules/opengl/buffer/framebufferobject.h [hunk 8]
```diff
@@ -36,12 +36,22 @@
...
 #include <vector>
+#include <optional>
 
 namespace inviwo {
 
+/**
+ * @brief Inviwo framebuffer wrapper.
+ *
+ * Handles creation and deletion of OpenGL framebuffer objects. Has functions for attachment and
+ * detachment of textures to the framebuffer. It Also keeps track of all attached texture ids. The
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/buffer/framebufferobject.h]: typos, 
"attachment"/"detachment" -> attaching/detaching 
"Also" -> "also"
"wapper" -> "wrapper"

File: modules/opengl/include/modules/opengl/buffer/framebufferobject.h (hunk 3 above)
Comment (petersteneteg) [reply modules/opengl/include/modules/opengl/buffer/framebufferobject.h]: true... will refactor more

File: modules/opengl/include/modules/opengl/buffer/framebufferobject.h [hunk 9]
```diff
@@ -50,163 +60,296 @@ class IVW_MODULE_OPENGL_API FrameBufferObject {
...
+     * @param layer index of the layer to attach
+     * @return the attachment point used (@c GL_COLOR_ATTACHMENT0, ... )
      */
-    GLenum attachColorTextureLayer(Texture2DArray* texture, int attachmentNumber, int layer,
-                                   bool attachFromRear = false, int forcedLocation = -1);
+    GLenum attachColorTextureLayer(Texture2DArray* texture, int attachmentNumber, int layer);
 
     /**
-     * For attaching a 3D Texture
+     * @brief Attach a 3D Texture to the framebuffer
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/buffer/framebufferobject.h]: maybe mention that "use `gl_Layer` in the geometry shader to set target z slice."

File: modules/opengl/include/modules/opengl/buffer/framebufferobject.h [hunk 10]
```diff
@@ -50,163 +60,296 @@ class IVW_MODULE_OPENGL_API FrameBufferObject {
...
     void attachTexture(Texture3D* texture, GLenum attachmentID);
+
+    /**
+     * @brief Attach a 3D Color Texture to the framebuffer.
+     *
+     * Will use the first available attachment point.
+     * @pre The framebuffer must be active.
+     * @param texture to attach
+     * @return The used attachment point (@c GL_COLOR_ATTACHMENT0, ... )
+     */
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/buffer/framebufferobject.h:200]: mention `gl_Layer`?

File: modules/opengl/include/modules/opengl/buffer/framebufferobject.h [hunk 11]
```diff
@@ -50,163 +60,296 @@ class IVW_MODULE_OPENGL_API FrameBufferObject {
...
+     * @param layer index of the layer to attach
      */
     void attachTextureLayer(Texture3D* texture, GLenum attachmentID, int layer);
+
+    /**
+     * @brief Attach a layer of a 3D Color Texture to the framebuffer.
+     * Will use the first available attachment point.
+     * @pre The framebuffer must be active.
+     * @param texture to attach
+     * @param layer index of the layer to attach
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/buffer/framebufferobject.h]: z layer?

File: modules/opengl/include/modules/opengl/buffer/framebufferobject.h [hunk 12]
```diff
@@ -50,163 +60,296 @@ class IVW_MODULE_OPENGL_API FrameBufferObject {
...
-                              int forcedLocation = -1);
+    GLenum attachColorTexture(Texture3D* texture, int attachmentNumber);
 
     /**
-     * For attaching a layer of a 3D Texture
+     * @brief Attach a layer of a 3D Texture to the framebuffer.
+     * @pre The framebuffer must be active.
+     * @param texture to attach
+     * @param attachmentID attachment point to use (@c GL_COLOR_ATTACHMENT0, ... )
+     * @param layer index of the layer to attach
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/buffer/framebufferobject.h]: z layer?

File: modules/opengl/include/modules/opengl/buffer/framebufferobject.h [hunk 13]
```diff
@@ -50,163 +60,296 @@ class IVW_MODULE_OPENGL_API FrameBufferObject {
...
      */
-    const std::vector<GLuint>& getDrawBuffersInUse() const { return attachedColorIds_; }
-    const GLenum* getDrawBuffersDeprecated() const;
-    int getMaxColorAttachments() const;
+    static int getMaxColorAttachments();
 
     /**
-     * Returns the location of the given attachment withing the registered draw buffers (e.g. used
-     * for glBindAttribLocation() and glFragDataLocation())
+     * @brief Check if there is any color attachments
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/buffer/framebufferobject.h]: either "is ... attachment" or "are ... attachments"

File: modules/opengl/include/modules/opengl/image/imagegl.h [hunk 14]
```diff
@@ -46,27 +45,84 @@ class Image;
 
 /**
  * \ingroup datastructures
+ * @brief OpenGL representation of a Image as a OpenGL framebuffer
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/image/imagegl.h]: typo "of _an_ Image"

File: modules/opengl/include/modules/opengl/image/imagegl.h [hunk 15]
```diff
@@ -46,27 +45,84 @@ class Image;
...
-
     virtual ImageGL* clone() const override;
 
-    void reAttachAllLayers(ImageType type = ImageType::AllLayers);
+    using ActiveState = std::tuple<utilgl::ActivateFBO, utilgl::GlBoolState, utilgl::DepthMaskState,
+                                   utilgl::ViewportState>;
 
-    void activateBuffer(ImageType type = ImageType::AllLayers);
+    /**
+     * @brief Store the current "state" and active the FBO.
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/image/imagegl.h]: "activate"

File: modules/opengl/include/modules/opengl/image/imagegl.h [hunk 16]
```diff
@@ -46,27 +45,84 @@ class Image;
...
     virtual size_t priority() const override;
 
+    /**
+     * @copydoc copyRepresentationsTo(ImageRepresentation* target) const
+     */
     bool copyRepresentationsTo(ImageGL* target) const;
-    bool updateFrom(const ImageGL*);
+
+    /**
+     * @brief Copies @p source into this using Blitting and PBOs if needed.
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/image/imagegl.h]: into this? suggestion "into this ImageGL" or "into this instance"

File: modules/opengl/include/modules/opengl/image/imagegl.h [hunk 17]
```diff
@@ -46,27 +45,84 @@ class Image;
...
+     *
+     * Does not care about aspect ratios, copies the source image into the this using "nearest"
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/image/imagegl.h]: typo "into the this"

File: modules/opengl/include/modules/opengl/image/imagegl.h [hunk 18]
```diff
@@ -46,27 +45,84 @@ class Image;
...
+     * interpolation. If some layers are not attached to the FBO they will be copied using PBO
+     * instead.
+     * @param source Image to copy from
+     * @return true
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/image/imagegl.h:124]: always?

**Review by ResearchDaniel — DISMISSED** [review]

**Review by martinfalk — DISMISSED** [review]
nice job with cleaning up the FBO states. Some typos...

File: modules/opengl/include/modules/opengl/image/imagegl.h (hunk 18 above)
Comment (petersteneteg) [reply modules/opengl/include/modules/opengl/image/imagegl.h:124]: yes

**Review by martinfalk — APPROVED** [review]

**Pull request merged**